core/src/main/java/hudson/tasks/LogRotator.java

```diff
@@ -148,9 +171,9 @@ public void perform(Job<?, ?> job) throws IOException, InterruptedException {
 
         LOGGER.log(FINE, "Running the log rotation for {0} with numToKeep={1} daysToKeep={2} artifactNumToKeep={3} artifactDaysToKeep={4}", new Object[] {job, numToKeep, daysToKeep, artifactNumToKeep, artifactDaysToKeep});
 
-        // always keep the last successful and the last stable builds
-        Run lsb = job.getLastSuccessfulBuild();
-        Run lstb = job.getLastStableBuild();
+        // if configured keep the last successful and the last stable builds
+        Run lsb = removeLastSuccessfulBuild ? null : job.getLastSuccessfulBuild();
+        Run lstb = removeLastStableBuild ? null : job.getLastStableBuild();
```

Comment:
hmm I think this logic may not be 100% correct.

I have selected "Remove last stable build" - Yes

and then I have a stable build (success) and 2 failures.

My stable build never gets removed because I have set Remove last successful build - no.

The UI seems to be a bit too tied to the implementation. 

At a minimum I think if remove last stable build is selected as yes and the last successful build is a stable build then it should be removed.

----

I wonder if we need both UI options, would it be enough to just add the stable option? 

Or do you have any ideas on how it could be simpler?

Reply:
Indeed, having a single UI setting is simpler.